### PR TITLE
backupccl: explicitly parse SHOW BACKUP options

### LINKED ISSUE #82912: backupccl: add `SHOW BACKUP ... with incremental_location` for locality aware backups
Currently, SHOW BACKUP works with the `incremental_location` parameter or for locality aware backups, but not for locality aware backups _with_ the `incremental_location` parameter. This requires renovating SHOW BACKUP's implementation in `sql.y` to look more like BACKUP or RESTORE's more flexible sql.y implementation. 

Jira issue: CRDB-16724

### PR DESCRIPTION
**Pull request by msbutler**
Fixes: https://github.com/cockroachdb/cockroach/issues/82912

Release note (sql change): Previously, SHOW BACKUP options would get parsed as
`kv_options`, which meant that a user could not pass multiple values to a show
backup option, causing feature gaps in SHOW BACKUP relative to BACKUP and
RESTORE. This patch rewrites the show backup option parser, closing the
following feature gaps:
 1. A user can now pass and check multiple KMS URIs in
SHOW BACKUP 
2. A user can pass locality aware incremental_locations, allowing a
user to also pass the check_files parameter to a locality aware backup chain
that also specifies the backup incremental location.

Note that while this patch introduces a couple new words to the CRDB SQL
syntax, the same SHOW BACKUP options should remain documented, specifically:
- [public option] -> value
- AS_JSON -> N/A
- CHECK_FILES -> N/A
- INCREMENTAL_LOCATION -> string, with potentially multiple uris
- DEBUG_IDS -> N/A
- KMS -> string, with potentially multiple uris
- PRIVILEGES -> N/A
- ENCRYPTION_PASSPHRASE -> string

### CODE REVIEW

**Comment by benbardin** [comment]
Oof. I see where this is coming from but adding so many keywords feels off, since this is a sub-option for one query. Why isn't this an issue for the other commands?

**Comment by msbutler** [comment]
> Why isn't this an issue for the other commands?

Like other cockroach sql cmds with an extensive Options list? I'm not sure I follow. I think what I propose is the conventional way to do it? E.g. backup, restore, copy

**Review by benbardin — COMMENTED** [review]
Discussed offline: Michael showed me that the base `BACKUP` and `RESTORE` commands already use this query structure, e.g. with a bunch of keywords. By adding the keywords here, we're bringing the SHOW BACKUP command to parity with these (both in query data structure, and in user functionality)

File: pkg/sql/sem/tree/show.go [hunk 1]
```diff
@@ -129,12 +131,182 @@ func (node *ShowBackup) Format(ctx *FmtCtx) {
...
+			return v1, errors.Newf("% option specified multiple times", label)
+		}
+		return v1, nil
+	}
+	return v2, nil
+}
+func combineStringOrPlaceholderOptList(
+	v1 StringOrPlaceholderOptList, v2 StringOrPlaceholderOptList, label string,
+) (StringOrPlaceholderOptList, error) {
+	if v1 != nil {
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/sql/sem/tree/show.go:250]: Yeah i think it is. It prevents the user from specifying INCREMENTAL_LOCATION multiple times, e.g.:

`SHOW BACKUP WITH INCREMENTAL_LOCATION=a, INCREMENTAL_LOCATION=b`

We don't allows this BACKUP and RESTORE either.

File: pkg/sql/sem/tree/show.go [hunk 2]
```diff
@@ -129,12 +131,182 @@ func (node *ShowBackup) Format(ctx *FmtCtx) {
...
+		o.DebugIDs == options.DebugIDs &&
+		cmp.Equal(o.IncrementalStorage, options.IncrementalStorage) &&
+		cmp.Equal(o.DecryptionKMSURI, options.DecryptionKMSURI) &&
+		o.EncryptionPassphrase == options.EncryptionPassphrase &&
+		o.Privileges == options.Privileges &&
+		o.DebugMetadataSST == options.DebugMetadataSST &&
+		o.EncryptionInfoDir == options.EncryptionInfoDir
+}
+
+func combineBools(v1 bool, v2 bool, label string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/sql/sem/tree/show.go:232]: nice idea. done.

File: pkg/sql/parser/sql.y [hunk 3]
```diff
@@ -15810,6 +15883,7 @@ unreserved_keyword:
 | ENCODING
 | ENCRYPTED
 | ENCRYPTION_PASSPHRASE
+| ENCRYPTION_INFO_DIR
```
Comment (stevendanna) [inline pkg/sql/parser/sql.y:15886]: Please add these to the `bare_label_keyword` list as well.

File: pkg/sql/sem/tree/show.go [hunk 4]
```diff
@@ -129,12 +131,179 @@ func (node *ShowBackup) Format(ctx *FmtCtx) {
...
 	}
 }
 
+type ShowBackupOptions struct {
+	AsJson               bool
+	CheckFiles           bool
+	DebugIDs             bool
+	IncrementalStorage   StringOrPlaceholderOptList
+	DecryptionKMSURI     StringOrPlaceholderOptList
+	EncryptionPassphrase Expr
```
Comment (stevendanna) [inline pkg/sql/sem/tree/show.go:146]: Just noting: We don't currently have an implementation of walkableStmt for ShowBackup. If we did, I would say we should probably be updating it to account for these new Exprs. We may want to add one in a follow up.

**Review by stevendanna — COMMENTED** [review]
I think my main question wrt all of the new keywords is whether we definitely want to promote some of these options into documented, supported options.

**Review by benbardin — APPROVED** [review]

File: pkg/sql/parser/sql.y (hunk 3 above)
Comment (msbutler) [reply pkg/sql/parser/sql.y:16006]: Done.

File: pkg/sql/sem/tree/show.go (hunk 4 above)
Comment (msbutler) [reply pkg/sql/sem/tree/show.go:148]: Sounds good. Just added a todo. I don't know enough about this method to understand why one would need it. It seems that no other SHOW cmd in this file implements one either.

**Review by msbutler — COMMENTED** [review]
@stevendanna on your question on keyword documentation: I tried to describe what needed to be documented in the commit message. Let me know if that satisfied you!

File: pkg/sql/parser/sql.y (hunk 3 above)
Comment (msbutler) [reply pkg/sql/parser/sql.y:16006]: also, you mentioned in some meeting today that you'd file/ or work on a stability period issue that would create a linter for syntax changes right? I imagine we also need to add a bunch of backup/restore keywords to this list as well.

**Comment by msbutler** [comment]
unrelated extended CI failure

**Comment by msbutler** [comment]
unrelated CI flake

**Comment by stevendanna** [comment]
>  I tried to describe what needed to be documented in the commit message. 

Heh, I read that, thought it was good, and then forgot it by the time I finished my review. Sorry about that! LGTM

**Comment by msbutler** [comment]
bors r=benbardin

**Pull request merged**